### ISSUE TITLE
Already referencing a newer version of 'packagename'

### ISSUE THREAD

**Issue by ferventcoder** [chocolatey/choco#227]
Originally referenced at https://github.com/scriptcs/scriptcs/issues/1015

This isn't exactly a bug, but the message might be. See https://nuget.codeplex.com/workitem/1549

Folks can get this when they have a package that limits another package referencing a newer version. 

Other references:
* https://github.com/NuGet/Home/issues/736
* https://github.com/TheCloudlessSky/Harbour.RedisSessionStateStore/issues/15#issuecomment-38304106

### Workaround

Try each of these one at a time and rerun the upgrade to see if it fixes the issue:

* Set Cache Location as it is more deterministic in cleaning up caches - `choco config set cacheLocation c:\ProgramData\choco-cache`
* Look in both `C:\ProgramData\chocolatey\lib\everything` and `C:\ProgramData\chocolatey\lib\chocolatey-core.extension to verify they only have one nupkg file (without a version in the name of the nupkg). If you find any others, delete them.
* Run `choco upgrade <pkgId> --ignore-dependencies`

**Comment by nestalk** [comment]
I just had the same error as scriptcs/scriptcs#1015. I noticed that as i uninstalled Nuget.CommandLine chocolatey uninstalled two versions, 2.8.2 and 2.8.5.

```
2015-04-09 09:43:12,988 [INFO ] - ============================================================
2015-04-09 09:43:13,038 [INFO ] - Chocolatey v0.9.9.4
2015-04-09 09:43:13,128 [INFO ] - Uninstalling the following packages:
2015-04-09 09:43:13,128 [INFO ] - Nuget.Commandline
2015-04-09 09:43:17,688 [INFO ] - 
NuGet.CommandLine v2.8.2
2015-04-09 09:43:18,258 [WARN ] - This is try 1/3. Retrying after 300 milliseconds.
 Error converted to warning:
 Cannot create a file when that file already exists.

2015-04-09 09:43:18,558 [WARN ] - This is try 2/3. Retrying after 400 milliseconds.
 Error converted to warning:
 Cannot create a file when that file already exists.

2015-04-09 09:43:18,958 [ERROR] - Maximum tries of 3 reached. Throwing error.
2015-04-09 09:43:18,968 [WARN ] - Move failed with message:
 Cannot create a file when that file already exists.

 Attempting backup move method.
2015-04-09 09:43:23,838 [INFO ] - 
NuGet.CommandLine v2.8.5
2015-04-09 09:43:23,878 [INFO ] -  Skipping auto uninstaller - No registry snapshot.
2015-04-09 09:43:24,478 [INFO ] -  NuGet.CommandLine has been successfully uninstalled.
2015-04-09 09:43:24,478 [WARN ] - 
Chocolatey uninstalled 2/2 packages. 0 packages failed.
```
In the logs in scriptcs/scriptcs#1015 it shows that he has nuget 2.8.3 installed but it is trying to uninstall 2.8.2. Could it be an issue with multiple versions of the same package installed at once?

I deleted the files that remained in the lib folder and then scriptcs would install.

**Comment by ferventcoder** [comment]
As a workaround, determine if you have multiple versions of nuget.commandline installed by typing 

`choco list -localonly -all`

**Comment by digital-carver** [comment]
I'm experiencing a simliar issue when upgrading LinkShellExtension, with `vcredist2005` in place of nuget.commandline (as commented [here](https://chocolatey.org/packages/LinkShellExtension#comment-2067014271) )


In this case, I'm vary of uninstalling all but the most recent versions too, since the general recommendation (eg: on SuperUser [here](http://superuser.com/a/242702/3200) and [here](http://superuser.com/questions/173403/why-are-there-so-many-different-visual-c-redistributables/619114#619114), on MS Community discussions [here](http://answers.microsoft.com/en-us/windows/forum/windows_7-windows_programs/multiple-microsoft-visual-c-redistributable/0608a19a-38a8-4885-9eb4-128a2bfad0a3) ). 

By the way, is there a way to find out which Chocolatey package depends upon each particular version of vcredist2005? I apparently have `vcredist2005 8.0.50727.763` and `vcredist2005 8.1.0.0` as per `clist -lo -all`.

**Comment by ferventcoder** [comment]
which one is actually installed on the machine @digital-carver ?

**Comment by chucker** [comment]
Had this as well.

Before, Link Shell Extension 3.8.6.1 (from the Chocolatey package) and vcredist2005 8.0.61000 und 8.0.56336 were installed.

'cup linkshellextension -y' failed with the above message. 'cinst vcredist2005' claimed it was already installed, even though the package lists its current version as 8.1.0.0 (this package metadata appears to be incorrect). 'cinst vcredist2005 --force' fixed it, even though Control Panel still lists the same versions.

This may be an arch issue, too. linkshellextension gets installed for both x86 and x64, but for x86, only vcredist2005 8.0.56336 is installed, whereas both versions are installed for x64.

**Comment by ferventcoder** [comment]
Thanks for the update.

**Comment by ferventcoder** [comment]
Also @timbrown5

**Comment by ferventcoder** [comment]
This could be a caching issue as well - https://github.com/TheCloudlessSky/Harbour.RedisSessionStateStore/issues/15#issuecomment-38304106

**Comment by ferventcoder** [comment]
https://chocolatey.org/docs/troubleshooting covers some of this as well.

**Comment by jberezanski-mdg** [comment]
Happens right now with [`vcredist-all`](https://chocolatey.org/packages/vcredist-all):

```
PS C:\> cinst -y vcredist-all
Chocolatey v0.10.11
Installing the following packages:
vcredist-all
By installing you accept licenses for the packages.
Progress: Downloading vcredist2005 8.1.0.20160118... 100%
Progress: Downloading vcredist-all 1.0.0... 100%
Progress: Downloading vcredist2008 9.0.30729.6163... 100%
Progress: Downloading vcredist2010 10.0.40219.2... 100%
Progress: Downloading vcredist2012 11.0.61031... 100%
Progress: Downloading vcredist2013 12.0.40660.20180427... 100%
Progress: Downloading vcredist140 14.20.27508.1... 100%
Progress: Downloading chocolatey-core.extension 1.3.3... 100%
Progress: Downloading KB3033929 1.0.4... 100%
Progress: Downloading chocolatey-windowsupdate.extension 1.0.3... 100%
Progress: Downloading KB3035131 1.0.2... 100%
Progress: Downloading KB2999226 1.0.20181019... 100%
Progress: Downloading vcredist2015 14.0.24215.20170201... 100%
Progress: Downloading vcredist2017 14.16.27027.1... 100%
Progress: Downloading vcredist140 14.16.27027.1... 100%
vcredist-all not installed. An error occurred during installation:
 Already referencing a newer version of 'vcredist140'.
vcredist-all package files install completed. Performing other installation steps.
The install of vcredist-all was NOT successful.
vcredist-all not installed. An error occurred during installation:
 Already referencing a newer version of 'vcredist140'.
```

Dependencies are:
- vcredist-all 1.0.0 -> [vcredist140](https://chocolatey.org/packages/vcredist140) (≥ 14.10.25008)
- vcredist-all 1.0.0 -> [vcredist2017](https://chocolatey.org/packages/vcredist2017) (≥ 14.10.25008)
- vcredist2017 14.16.27027.1 -> [vcredist140](https://chocolatey.org/packages/vcredist140) (= 14.16.27027.1)
and vcredist140 has newer, approved versions (14.16.27027.20190414, 14.20.27508.1).

Installing explicitly vcredist2017 and then vcredist-all works fine.

(The issue might no longer reproduce once vcredist2017 version 14.16.27027.20190415, which changes the dependency from "=" to "<=", is approved.)

**Comment by jberezanski** [comment]
Happens [right now](https://chocolatey.org/packages/vcredist-all#comment-4423614284) with [`vcredist-all`](https://chocolatey.org/packages/vcredist-all):

```
PS C:\> cinst -y vcredist-all
Chocolatey v0.10.11
Installing the following packages:
vcredist-all
By installing you accept licenses for the packages.
Progress: Downloading vcredist2005 8.1.0.20160118... 100%
Progress: Downloading vcredist-all 1.0.0... 100%
Progress: Downloading vcredist2008 9.0.30729.6163... 100%
Progress: Downloading vcredist2010 10.0.40219.2... 100%
Progress: Downloading vcredist2012 11.0.61031... 100%
Progress: Downloading vcredist2013 12.0.40660.20180427... 100%
Progress: Downloading vcredist140 14.20.27508.1... 100%
Progress: Downloading chocolatey-core.extension 1.3.3... 100%
Progress: Downloading KB3033929 1.0.4... 100%
Progress: Downloading chocolatey-windowsupdate.extension 1.0.3... 100%
Progress: Downloading KB3035131 1.0.2... 100%
Progress: Downloading KB2999226 1.0.20181019... 100%
Progress: Downloading vcredist2015 14.0.24215.20170201... 100%
Progress: Downloading vcredist2017 14.16.27027.1... 100%
Progress: Downloading vcredist140 14.16.27027.1... 100%
vcredist-all not installed. An error occurred during installation:
 Already referencing a newer version of 'vcredist140'.
vcredist-all package files install completed. Performing other installation steps.
The install of vcredist-all was NOT successful.
vcredist-all not installed. An error occurred during installation:
 Already referencing a newer version of 'vcredist140'.
```

Dependencies are:
- vcredist-all 1.0.0 -> [vcredist140](https://chocolatey.org/packages/vcredist140) (≥ 14.10.25008)
- vcredist-all 1.0.0 -> [vcredist2017](https://chocolatey.org/packages/vcredist2017) (≥ 14.10.25008)
- vcredist2017 14.16.27027.1 -> [vcredist140](https://chocolatey.org/packages/vcredist140) (= 14.16.27027.1)
and vcredist140 has newer, approved versions (14.16.27027.20190414, 14.20.27508.1).

Installing explicitly vcredist2017 and then vcredist-all works fine.

(The issue might no longer reproduce once vcredist2017 version 14.16.27027.20190415, which changes the dependency from "=" to "<=", is approved.)

**Comment by ferventcoder** [comment]
@jberezanski this is expected behavior - I dive into this in more detail at https://github.com/chocolatey/ChocolateyGUI/issues/632#issuecomment-483299453. This is a reason you can't have a single package to depend on that would install multiple versions of something.

**Comment by flcdrg** [comment]
I'm seeing this with packages that depend on `chocolatey-core.extension`.

Looking in `$env:ChocolateyInstall\lib` I see both `chocolatey-core.extension` and `chocolatey-core.extension.1.3.3`

The first appears to be version 1.3.4. Manually deleting the `chocolatey-core.extension.1.3.3` folder fixed this (and allowed me to upgrade other packages that have a dependency on chocolatey-core.extension again.

**Comment by brimwats** [comment]
I am also having this issue

> You have f.lux v4.100 installed. Version 4.111 is available based on your source(s).
> f.lux not upgraded. An error occurred during installation:
>  Already referencing a newer version of 'chocolatey'.
> f.lux package files upgrade completed. Performing other installation steps.
> The upgrade of f.lux was NOT successful.
> f.lux not upgraded. An error occurred during installation:
>  Already referencing a newer version of 'chocolatey'.
> You have f.lux.install v4.100 installed. Version 4.111 is available based on your source(s).
> f.lux.install not upgraded. An error occurred during installation:
>  Already referencing a newer version of 'chocolatey'.
> f.lux.install package files upgrade completed. Performing other installation steps.
> The upgrade of f.lux.install was NOT successful.
> f.lux.install not upgraded. An error occurred during installation:
>  Already referencing a newer version of 'chocolatey'.
> You have git v2.21.0 installed. Version 2.24.0 is available based on your source(s).
> git not upgraded. An error occurred during installation:
>  Already referencing a newer version of 'chocolatey'.
> git package files upgrade completed. Performing other installation steps.
> The upgrade of git was NOT successful.
> git not upgraded. An error occurred during installation:
>  Already referencing a newer version of 'chocolatey'.
> You have git.install v2.21.0 installed. Version 2.24.0.2 is available based on your source(s).
> git.install not upgraded. An error occurred during installation:
>  Already referencing a newer version of 'chocolatey'.
> git.install package files upgrade completed. Performing other installation steps.
> The upgrade of git.install was NOT successful.
> git.install not upgraded. An error occurred during installation:
>  Already referencing a newer version of 'chocolatey'.


I've not found an extra chocolatey-core.extension where everyone else has, but I did find one in 

C:\Users\USER\AppData\Local\NuGet\Cache

Deleting didn't seem to do anything. I'm a bit wary of running the command to ignore dependencies--will it bork anything?

**Comment by charltonstanley** [comment]
>Manually deleting the `chocolatey-core.extension.1.3.3` folder fixed this (and allowed me to upgrade other packages that have a dependency on chocolatey-core.extension again.

this worked perfectly for me. thank you!

**Comment by brimwats** [comment]
I went ahead and tried to delete chocolatey-core.extension.1.3.3 -- no luck

You have f.lux v4.100 installed. Version 4.111 is available based on your source(s).
f.lux not upgraded. An error occurred during installation:
 Already referencing a newer version of 'chocolatey'.
f.lux package files upgrade completed. Performing other installation steps.
The upgrade of f.lux was NOT successful.
f.lux not upgraded. An error occurred during installation:
 Already referencing a newer version of 'chocolatey'.

> You have f.lux.install v4.100 installed. Version 4.111 is available based on your source(s).
> f.lux.install not upgraded. An error occurred during installation:
>  Already referencing a newer version of 'chocolatey'.
> f.lux.install package files upgrade completed. Performing other installation steps.
> The upgrade of f.lux.install was NOT successful.
> f.lux.install not upgraded. An error occurred during installation:
>  Already referencing a newer version of 'chocolatey'.

& etc

**Comment by flcdrg** [comment]
@brimwats yeah it may not be just that package. In my case I had a few where there were duplicates (or old versions hanging around) that once I manually removed the folders, it unblocked upgrading.

In your case, I'd see if you have multiple `chocolatey` folders - with one probably having a version number suffix.

**Comment by brimwats** [comment]
@flcdrg would you mind telling me what you think is safe to delete? I generally try to be cautious about rampant deletion of program files 
![image](https://user-images.githubusercontent.com/49103046/69505656-031b7c80-0ef9-11ea-9171-294f8f636850.png)


![image](https://user-images.githubusercontent.com/49103046/69505666-0d3d7b00-0ef9-11ea-8013-d0b4dbf5df90.png)


![image](https://user-images.githubusercontent.com/49103046/69505678-14fd1f80-0ef9-11ea-81ad-276a92d09ed9.png)


![image](https://user-images.githubusercontent.com/49103046/69505685-1e868780-0ef9-11ea-84d9-62d39ed4048a.png)

**Comment by brimwats** [comment]
(there are about ~7 more pages of .ps1 items after this), I figured they were not relavant)

**Comment by flcdrg** [comment]
Just look in `C:\ProgramData\chocolatey\lib` and take a closer look where you have 'duplicates' (but only where you know you didn't intentionally install additional versions using the `-m` parameter).

eg. For me, I can see one that looks a bit odd:
![image](https://user-images.githubusercontent.com/384747/69505873-00ae3680-0f7c-11ea-91d6-0a75eafcadaa.png)

Looking closer at the contents of those directories, the 'unversioned' folder seems to actually also contain 1.0.4 (comparing the .nuspec file contents), so I think I'm safe deleting `chocolatey-windowsupdate.extension.1.0.4`

I'm not concerned about all the dotnetcore-sdk.2.1.nnn folders because I intentionally installed those with `-m` to allow multiple versions of those packages.

**Comment by brimwats** [comment]
No luck with this. However I just took the steps of fully uninstalling and then deleting files manually, then reinstalling flux, git and python which seemed to be fighting me. This seemed to fix it, then I ran cup -all -y again and only got problems with 
Failures
 - git.install (exited 1) - git.install not upgraded. An error occurred during installation:
 Already referencing a newer version of 'chocolatey'.

but it's a matter of time before it contines with other packages, ugh

**Comment by flcdrg** [comment]
So out of interest, do you have more than one `chocolatey` folder in `C:\ProgramData\chocolatey\lib` ?

**Comment by brimwats** [comment]
Not really? I'm unsure if the extensions, core, dotnet or windows counts as more than one. here's all I have:
![image](https://user-images.githubusercontent.com/49103046/69511846-b7280200-0f0f-11ea-8aed-f8b96f8922f1.png)

**Comment by flcdrg** [comment]
Ok.. just the one in that folder then.

Also take a look at https://chocolatey.org/docs/troubleshooting#already-referencing-a-newer-version-of-packagename (as pointed to by Rob earlier) in case there's some stray .nupkg files hanging around

**Comment by brimwats** [comment]
I ran these earlier, but didn't get too much:

> PS C:\WINDOWS\system32> gci -Path "$env:ChocolateyInstall\lib" -Recurse -Filter "*.nupkg" | %{ $_.FullName }
> C:\ProgramData\chocolatey\lib\7zip\7zip.nupkg
> C:\ProgramData\chocolatey\lib\7zip.install\7zip.install.nupkg
> C:\ProgramData\chocolatey\lib\adobe-creative-cloud\adobe-creative-cloud.nupkg
> C:\ProgramData\chocolatey\lib\audacity\audacity.nupkg
> C:\ProgramData\chocolatey\lib\AutoRuns\AutoRuns.nupkg
> C:\ProgramData\chocolatey\lib\calibre\calibre.nupkg
> C:\ProgramData\chocolatey\lib\chocolatey\chocolatey.nupkg
> C:\ProgramData\chocolatey\lib\chocolatey-core.extension\chocolatey-core.extension.nupkg
> C:\ProgramData\chocolatey\lib\chocolatey-dotnetfx.extension\chocolatey-dotnetfx.extension.nupkg
> C:\ProgramData\chocolatey\lib\chocolatey-windowsupdate.extension\chocolatey-windowsupdate.extension.nupkg
> C:\ProgramData\chocolatey\lib\ChocolateyGUI\ChocolateyGUI.nupkg
> C:\ProgramData\chocolatey\lib\curl\curl.nupkg
> C:\ProgramData\chocolatey\lib\disable-nvidia-telemetry\disable-nvidia-telemetry.nupkg
> C:\ProgramData\chocolatey\lib\ditto\ditto.nupkg
> C:\ProgramData\chocolatey\lib\dotnet4.6.2\dotnet4.6.2.nupkg
> C:\ProgramData\chocolatey\lib\dotnetfx\dotnetfx.nupkg
> C:\ProgramData\chocolatey\lib\dropbox\dropbox.nupkg
> C:\ProgramData\chocolatey\lib\f.lux\f.lux.nupkg
> C:\ProgramData\chocolatey\lib\f.lux.install\f.lux.install.nupkg
> C:\ProgramData\chocolatey\lib\filezilla\filezilla.nupkg
> C:\ProgramData\chocolatey\lib\foobar2000\foobar2000.nupkg
> C:\ProgramData\chocolatey\lib\git\git.nupkg
> C:\ProgramData\chocolatey\lib\git.install\git.install.nupkg
> C:\ProgramData\chocolatey\lib\github-desktop\github-desktop.nupkg
> C:\ProgramData\chocolatey\lib\imagemagick.app\imagemagick.app.nupkg
> C:\ProgramData\chocolatey\lib\jre8\jre8.nupkg
> C:\ProgramData\chocolatey\lib\KB2919355\KB2919355.nupkg
> C:\ProgramData\chocolatey\lib\KB2919442\KB2919442.nupkg
> C:\ProgramData\chocolatey\lib\KB2999226\KB2999226.nupkg
> C:\ProgramData\chocolatey\lib\KB3033929\KB3033929.nupkg
> C:\ProgramData\chocolatey\lib\KB3035131\KB3035131.nupkg
> C:\ProgramData\chocolatey\lib\keypirinha\keypirinha.nupkg
> C:\ProgramData\chocolatey\lib\kubernetes-cli\kubernetes-cli.nupkg
> C:\ProgramData\chocolatey\lib\Minikube\Minikube.nupkg
> C:\ProgramData\chocolatey\lib\mm-choco.extension\mm-choco.extension.nupkg
> C:\ProgramData\chocolatey\lib\netfx-4.6.2\netfx-4.6.2.nupkg
> C:\ProgramData\chocolatey\lib\nodejs-lts\nodejs-lts.nupkg
> C:\ProgramData\chocolatey\lib\notepadplusplus\notepadplusplus.nupkg
> C:\ProgramData\chocolatey\lib\notepadplusplus.install\notepadplusplus.install.nupkg
> C:\ProgramData\chocolatey\lib\pdfsam\pdfsam.nupkg
> C:\ProgramData\chocolatey\lib\pdfsam.install\pdfsam.install.nupkg
> C:\ProgramData\chocolatey\lib\pip\pip.nupkg
> C:\ProgramData\chocolatey\lib\PowerShellHere\PowerShellHere.nupkg
> C:\ProgramData\chocolatey\lib\python3\python3.nupkg
> C:\ProgramData\chocolatey\lib\qbittorrent\qbittorrent.nupkg
> C:\ProgramData\chocolatey\lib\quicklook\quicklook.nupkg
> C:\ProgramData\chocolatey\lib\sqlitebrowser\sqlitebrowser.nupkg
> C:\ProgramData\chocolatey\lib\sqlitebrowser.install\sqlitebrowser.install.nupkg
> C:\ProgramData\chocolatey\lib\treesizefree\treesizefree.nupkg
> C:\ProgramData\chocolatey\lib\typora\typora.nupkg
> C:\ProgramData\chocolatey\lib\vagrant\vagrant.nupkg
> C:\ProgramData\chocolatey\lib\vcredist140\vcredist140.nupkg
> C:\ProgramData\chocolatey\lib\vcredist2010\vcredist2010.nupkg
> C:\ProgramData\chocolatey\lib\vcredist2013\vcredist2013.nupkg
> C:\ProgramData\chocolatey\lib\vcredist2015\vcredist2015.nupkg
> C:\ProgramData\chocolatey\lib\virtualbox\virtualbox.nupkg
> C:\ProgramData\chocolatey\lib\vlc\vlc.nupkg
> C:\ProgramData\chocolatey\lib\Wget\Wget.nupkg
> C:\ProgramData\chocolatey\lib\winscp\winscp.nupkg
> C:\ProgramData\chocolatey\lib\winscp.install\winscp.install.nupkg
> C:\ProgramData\chocolatey\lib\youtube-dl-gui\youtube-dl-gui.nupkg
> C:\ProgramData\chocolatey\lib\youtube-dl-gui.install\youtube-dl-gui.install.nupkg
> C:\ProgramData\chocolatey\lib\youtube-mp3\youtube-mp3.nupkg

The second script doesn't return anything.

**Comment by brimwats** [comment]
Has anyone had any luck running updates since this was made?

**Comment by rikkit** [comment]
I got this error when running `cup git`.

I ran `choco list -localonly -all` as suggested in this thread and it reported that I had three versions of chocolatey installed. 

I ran `choco uninstall chocolatey -v 0.0.9.4` and it prompted me for which chocolatey version to uninstall. I chose 0.9.9.4. The command failed, I re-ran  `choco list -localonly -all` and there was only one instance of chocolatey.

Rerunning `cup git` worked as expected.

Not sure what was going on here, but hope this helps someone.

**Comment by starrychloe** [comment]
This ruby2.devkit package fails to upgrade with this error.
<pre>
C:\Users\Chloe>choco upgrade ruby2.devkit
Chocolatey v0.10.15
Upgrading the following packages:
ruby2.devkit
By upgrading you accept licenses for the packages.

You have ruby2.devkit v4.7.2.2013022402 installed. Version 4.7.2.2013022403 is available based on your source(s).
Progress: Downloading 7zip.commandline 16.02.0.20170209... 100%
Progress: Downloading 7zip.portable 19.0... 100%
Progress: Downloading ruby2.devkit 4.7.2.2013022403... 100%
ruby2.devkit not upgraded. An error occurred during installation:
 Already referencing a newer version of 'chocolatey-core.extension'.
ruby2.devkit package files upgrade completed. Performing other installation steps.
The upgrade of ruby2.devkit was NOT successful.
ruby2.devkit not upgraded. An error occurred during installation:
 Already referencing a newer version of 'chocolatey-core.extension'.
 Unsuccessful operation for ruby2.devkit.
  Rollback to previous version (package files only)?([Y]es/[N]o): y


Chocolatey upgraded 0/1 packages. 1 packages failed.
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).

Failures
 - ruby2.devkit (exited 1) - ruby2.devkit not upgraded. An error occurred during installation:
 Already referencing a newer version of 'chocolatey-core.extension'.
</pre>

**Comment by easybe** [comment]
> This isn't exactly a bug, but the message might be.

I disagree, see: https://github.com/chocolatey/choco/issues/2143#issuecomment-905649744

**Comment by asheroto** [comment]
> Originally referenced at [scriptcs/scriptcs#1015](https://github.com/scriptcs/scriptcs/issues/1015)
>
> ### Workaround
> Try each of these one at a time and rerun the upgrade to see if it fixes the issue:
> 
> * Look in both `C:\ProgramData\chocolatey\lib\everything` and `C:\ProgramData\chocolatey\lib\chocolatey-core.extension to verify they only have one nupkg file (without a version in the name of the nupkg). If you find any others, delete them.



This fixed it for me.

**Comment by asheroto** [comment]
> ```powershell
>  Unable to uninstall 'chocolatey 1.2.1' because 'chocolateygui 1.1.1, php 7.4.20, msys2 20230127.0.0, nirlauncher 1.23.66, php 8.2.4, php 8.1.17' depend on it.
> ```

I may be wrong, but I don't think you can uninstall Chocolatey using Chocolatey. If you want to try it anyway, you can uninstall these packages first and then uninstall, or remove them from the lib folder, but check out this article first...

Here's the article regarding uninstall. There is a script at the bottom of the page if needed.

https://docs.chocolatey.org/en-us/choco/uninstallation

**Comment by corbob** [comment]
Looking at the `php` package, it takes a dependency on `chocolatey`. I think looking at the output @aetoni has provided that their issue is from having installed the `php` package side by side which cascaded to do a side by side install all of the dependencies (chocolatey 0.10.0 or newer is listed as a dependency on that package).

**WARNING: The below is just guesses, and is offered with no warranty or guarantees. It is highly recommended that you ensure appropriate backups before trying anything suggested**

You're probably safe removing the `$env:ChocolateyInstall/lib/chocolatey.1.2.1` directory which will solve the issue of Chocolatey blocking things. But you've still got side by side installs that will need to be handled appropriately (many of them may be able to be removed in similar ways, but the `php` packages are probably going to need some manual finagling).

**Comment by aetonsi** [comment]
Hi and thanks everyone for the replies.
Yes, i did install php `--side-by-side` because i need multiple versions of it. 
I deleted `$env:ChocolateyInstall/lib/chocolatey.1.2.1` and then the upgrade worked (or so it seems).
@corbob what kind of manual fiddling are you referring to?

**Comment by asheroto** [comment]
I had this problem happen again. I have `cup all` run every day and log the output. Today I noticed that 39/192 packages were failing to install because of this issue.

To fix, I went into `$env:ChocolateyInstall/lib` and only deleted folders starting with `chocolatey.` and ending with a number. For example, `chocolatey.1.1.0`, `chocolatey-windowsupdate.extension.1.0.4`, etc.

I deleted these folders (yours may be different versions)...

![image](https://user-images.githubusercontent.com/49938263/231534682-4580d3bf-13ad-4810-9950-7ee5996b7b27.png)

I kept all of the other folders.

After doing this and running `cup all` everything updated successfully.

**Comment by aetonsi** [comment]
Guys it happened again 😞 
chocolatey upgraded itself to 1.4, leaving a `chocolatey.1.3.1` folder behind.
`php` and `chocolateygui`, once again, couldn't be updated.
Ironically enough, `winget` was able to update chocolateygui... still, `choco outdated` kept seeing it as being outdated.

I deleted the choco 1.3.1 folder and everything got upgraded without problems.

**Comment by pauby** [comment]
If you have any side-by-side installs, then these are the kind of issues you are going to run into. It's one of the reasons we have removed it in 2.0.0.

> will they be present in choco v2 too

Side-by-side installs have been removed from v2.0.0.

> still, choco outdated kept seeing it as being outdated.

Winget upgraded the software, not the Chocolatey package. So Chocolatey will continue to see it as outdated.

**Comment by aetonsi** [comment]
> If you have any side-by-side installs, then these are the kind of issues you are going to run into. It's one of the reasons we have removed it in 2.0.0.

I understand `php`, i installed multiple versions of it by myself. But about chocolatey GUI, is it still a matter of side by side installation? i don't have multiple versions of it

> Side-by-side installs have been removed from v2.0.0.

Yes i know that, but i'd like to ask. Will there still be some way to install multiple versions of packages together on the same machine? i need multiple instances of php to test things, will i be able to have them?

> Winget upgraded the software, not the Chocolatey package. So Chocolatey will continue to see it as outdated.

Yes i understand that, but sometimes, with other packages i can't recall, when winget updated them, `choco upgrade all` still would try to update them (correctly), but if i run a `choco outdated` **before** the upgrade command, the "external" version change would be detected.
But apparently it happens only with a few packages, not all of 'em?

thank you for your time btw

**Comment by Jackenmen** [comment]
> Yes i know that, but i'd like to ask. Will there still be some way to install multiple versions of packages together on the same machine? i need multiple instances of php to test things, will i be able to have them?

The way this is solved for Python is to have separate packages for each major version, e.g. `python38`, `python39`, and so on. I imagine such a split would need to be requested for the PHP package as well.

**Comment by pauby** [comment]
> But about chocolatey GUI, is it still a matter of side by side installation? i don't have multiple versions of it

I don't understand that question. If you didn't install `chocolateygui`, or any package _depending_ on `chocolateygui` in a side-by-side way, then it should not have multiple versions.

> Will there still be some way to install multiple versions of packages together on the same machine?

Not in v2.0.0. We currently intend to add functionality that allows multiple versions to be installed in the future, but that is still very early to talk much about it. 

>  i need multiple instances of php to test things, will i be able to have them?

See the [docs](https://docs.chocolatey.org/en-us/create/create-packages#naming-packages-to-allow-for-side-by-side-installation) for more about this but it works pretty much as @Jackenmen wrote above.

> Yes i understand that, but sometimes, with other packages i can't recall, when winget updated them, choco upgrade all still would try to update them (correctly), but if i run a choco outdated before the upgrade command, the "external" version change would be detected.
But apparently it happens only with a few packages, not all of 'em?

I'm not sure that I understand this one. If you haven't upgraded a **package**, and it has a newer version than you had installed, it will be shown as outdated. If you are seeing something else, then I'd need further information on that before I could comment further.

**Comment by aetonsi** [comment]
> I don't understand that question. If you didn't install `chocolateygui`, or any package _depending_ on `chocolateygui` in a side-by-side way, then it should not have multiple versions.

`chocolateygui`, I installed it normally, i wouldn't need multiple versions of that, it's just a GUI for chocolatey. And for the same reason, i highly doubt that any other package would side-by-side install it, ever?

I need multiple versions of php instead, so i do have personally, manually installed them via `--allow-multiple`.
I think it's the only software i have with multiple versions on my pc, at all. Also it's not really "installed", it's just folders in `c:\tools`.
Still, when this `already referencing a newer version` thing happens, it happens with any package, not only with the ones `--sxs` installed.

In my first post in this thread i also had the problem with `msys2`, for which i don't need multiple versions, at all. I think `msys2` is the one requiring `chocolatey`, so shouldn't it be the other way around? is there a choco command to show a dependency tree like [johnnydep](https://pypi.org/project/johnnydep/) or [pipdeptree](https://pypi.org/project/pipdeptree/) for python/pip?

---

> > i need multiple instances of php to test things, will i be able to have them?
> 
> See the [docs](https://docs.chocolatey.org/en-us/create/create-packages#naming-packages-to-allow-for-side-by-side-installation) for more about this but it works pretty much as @Jackenmen wrote above.

What would be the appropriate way or place to give you guys feedback and an idea about this topic (if possible)? I don't want to flood here too much, as i'm already doing (sorry).

---

> I'm not sure that I understand this one. If you haven't upgraded a **package**, and it has a newer version than you had installed, it will be shown as outdated. If you are seeing something else, then I'd need further information on that before I could comment further.

I'm not sure if i expressed myself correctly, sorry. What happened was:
- running `choco upgrade all`, some package `X` i can't recall (it happened with different packages, over time), didn't update. Hash mismatch, remote file not found, whatever
- Then, running `winget upgrade --all`/`winget upgrade X`, the _software_ got updated.
- Then, running `choco upgrade all` again, it would still try to update `X`. So i canceled it as soon as it showed up as out of date during the `upgrade all` wall of text, to stop it from possibly messing up the software itself
- Then, out of curiosity i run `choco outdated`: the software/package `X` was no more out of date.
- Running `choco upgrade all`, `X` got skipped because it was not seen as out of date anymore

That's why i thought that `choco outdated` did some version checks or something..
Might it be a case of me canceling (CTLR+C) the upgrade too late, and choco marking the package as updated even though it didn't do anything?
Next time it happens i'll give you more details about it, in a dedicated issue. Sorry if i'm not very helpful right now.
